src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2127 @@
+                device=hidden_states.device,
+                dtype=hidden_states.dtype,
+                requires_grad=hidden_states.requires_grad,
+            )
+            return final_hidden_states, router_logits
+
+        # in this case we have multiple experts and need to do routing
+        hidden_states = hidden_states.view(-1, hidden_dim)
+        # router_logits: (batch * sequence_length, n_experts)
+        router_logits = self.router(hidden_states)
+        routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+        routing_weights, selected_experts = torch.topk(routing_weights, self.top_k, dim=-1)
+        # we cast back to the input dtype
+        routing_weights = routing_weights.to(hidden_states.dtype)
+
+        final_hidden_states = torch.zeros(
+            (batch_size * sequence_length, hidden_dim), dtype=hidden_states.dtype, device=hidden_states.device
+        )
+
+        # One hot encode the selected experts to create an expert mask
+        # this will be used to easily index which expert is going to be sollicitated
+        expert_mask = torch.nn.functional.one_hot(selected_experts, num_classes=self.num_experts).permute(2, 1, 0)
+
+        # Loop over all available experts in the model and perform the computation on each expert
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
+            idx, top_x = torch.where(expert_mask[expert_idx])
+
+            if top_x.shape[0] == 0:
+                continue
+
+            # in torch it is faster to index using lists than torch tensors
+            top_x_list = top_x.tolist()
+            idx_list = idx.tolist()
+
+            # Index the correct hidden states and compute the expert hidden state for
+            # the current expert. We need to make sure to multiply the output hidden
+            # states by `routing_weights` on the corresponding tokens (top-1 and top-2)
+            current_state = hidden_states[None, top_x_list].reshape(-1, hidden_dim)
+            current_hidden_states = expert_layer(current_state) * routing_weights[top_x_list, idx_list, None]
+
+            # However `index_add_` only support torch tensors for indexing so we'll use
+            # the `top_x` tensor here.
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+        final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
+        return final_hidden_states, router_logits
+
+
+class JambaAttentionDecoderLayer(nn.Module):
+    def __init__(self, config: JambaConfig, num_experts: int, layer_idx: int):
+        super().__init__()
+
+        self.self_attn = JAMBA_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+        num_experts_per_tok = config.num_experts_per_tok if num_experts > 1 else 1
+        self.moe = JambaSparseMoeBlock(config, num_experts=num_experts, num_experts_per_tok=num_experts_per_tok)
```

Comment:
```suggestion
        if num_experts:
            self.mlp = JambaSparseMoeBlock(config, num_experts=num_experts, num_experts_per_tok=num_experts_per_tok)
       else:
           self.mlp = JambaMlp()
```
this makes a bit more sense IMO, as the class will be effectively different representing different code paths!

Reply:
I'm with you, but maybe we can find a better name for `JambaSparseMoeBlock`/`JambaMLP` than `mlp`? I'm not a big fan of `mlp` because it implies a single MLP where actually we can have experts. 
So for non experts layers we'll have:
* `mlp.gate_proj`
* `mlp.up_proj`
* `mlp.down_proj`

And for expert layers we'll have:
* `mlp.router`
* `mlp.experts.0.gate_proj`
* `mlp.experts.0.up_proj`
* `mlp.experts.0.down_proj`
* `mlp.experts.1.gate_proj`
* `mlp.experts.1.up_proj`
* `mlp.experts.1.down_proj`
...

I think the fact that both start with `mlp` is confusing.. WDYT?

Reply:
Also, if we use different classes for expert / non-expert layers, we'll have to get the router_logits for the non-expert layers from somewhere. I see that in switch_transformers it's just None for non-expert layers, and the code handling that is in the caller function - forward of `SwitchTransformersLayerFF`. 
We can do something similar here and introduce a `JambaFeedForwardLayer` that has a `JambaMLP` or a `JambaSparseMoeBlock` depending on `num_experts`. WDYT?

Reply:
OK I think I found some sort of solution:
* Each of `JambaAttentionDecoderLayer` and `JambaMambaDecoderLayer` have a `self.feed_forward` attribute.
* It can be a `JambaMLP` or a `JambaSparseMoeBlock`, depending if this is an expert layer or not. `JambaSparseMoeBlock` returns the router logits as well as the hidden states. `JambaMLP` returns only the hidden states.
* Accordingly, `pre_moe_layernorm` changed to `pre_ff_layernorm`.
* `JambaMLP` is now copied from Mistral
* `JambaSparseMoeBlock` cannot be copied from Mixtral since Mixtral do the routing weights normalization which we don't (as explained in [this comment](https://github.com/huggingface/transformers/pull/29943#discussion_r1545723627)).
* In `JambaAttentionDecoderLayer`/`JambaMambaDecoderLayer`, we check if the return value from `self.feed_forward` is a tuple or not (similar to what happens in `SwitchTransformersLayerFF`). If it's a tuple, we're an expert layer and the second item is the router logits. If not, we're a non-expert layer and we set the router logits to `None`.
* Similar to what we now do with the attentions, we don't append `None` router logits to the overall router logits returned from `JambaModel` (so `outputs.router_logits` has length shorter than `num_hidden_layers`)
* Thus, `load_balancing_loss_func` is now copied from Mixtral
* `JambaAttentionDecoderLayer` cannot be copied from Mixtral due to the special treatment with the `None` router logits